### ISSUE TITLE
Switching Unauthenticated Users to Authenticated Users

### ISSUE THREAD

**Issue by ahmetcetin** [aws-amplify/amplify-js#2089]
It's common scenario, user starts as unauthenticated, later registers. I allow user to use services as unauthenticated in identity pool. Using amplify, if I run this scenario, an identity created as unauthenticated, later user signs up, unauthenticated identity stays as it is, and new authenticated identity created. It looks there is a way to convert unauthenticated users to authenticated ones in cognito, any idea how to achieve it using amplify?

Initially Unauthenticated User
```
// set the default config object
var creds = new AWS.CognitoIdentityCredentials({
    IdentityPoolId: 'us-east-1:1699ebc0-7900-4099-b910-2df94f52a030'
});
AWS.config.credentials = creds;
``` 

Switch to Authenticated User
```
// Called when an identity provider has a token for a logged in user
function userLoggedIn(providerName, token) {
    creds.params.Logins = creds.params.Logins || {};
    creds.params.Logins[providerName] = token;

    // Expire credentials to refresh them on the next request
    creds.expired = true;
}
```
Do we have access to CognitoIdentityCredentials function through amplify?

** Which Category is your question related to? **
Auth
** What AWS Services are you utilizing? **
S3, AppSync
** Provide additional details e.g. code snippets **
https://docs.aws.amazon.com/cognito/latest/developerguide/switching-identities.html

**Comment by elorzafe** [comment]
Hi @ahmetcetin thanks for your feedback, what is the use you have in mind?

**Comment by jarstingstall** [comment]
👍

**Comment by marwanhilmi** [comment]
Am looking for support for this feature as well, currently using `amazon-cognito-identity-js` directly but would like to switch to `Amplify.Auth`. Can we possibly workaround by manually expiring the credentials and forcing a refresh?

With `amazon-cognito-identity-js`, am currently doing `AWS.config.credentials.expired = true` before fetching new authenticated credentials.

**Comment by salimhamed** [comment]
@elorzafe What would you suggest as a work around?

**Comment by NikoMontana** [comment]
I cannot even let unauth users use api, from switching between unauth. and auth. are different worlds. Did anybody solve this?

**Comment by powerful23** [comment]
@salimhamed sorry to be late. I am currently working this issue. The code you posted should work as when `Auth.signIn()` returns, Amplify will get an authenticated credentials in the background. Can you make sure you are using the latest version? Also you can turn on the debug mode: `window.LOG_LEVEL='DEBUG'` in your app to get the debug info.

**Comment by powerful23** [comment]
@ahmetcetin to make this issue clear, the feature wanted is to transfer the unauthenticated identity id into the authenticated one at the first time when the user signs in. Because as I know you cannot assign the unauthenticated identity id to a user who already has an authenticated one. So this case can only happen when the user first signs in.

**Comment by ahmetcetin** [comment]
@powerful23 right, it's exactly as you describe.

**Comment by salimhamed** [comment]
@powerful23 Thanks for the reply!  Yes, I'll make sure I'm on the latest version and turn log_level to debug, then report back.

**Comment by ajhool** [comment]
To provide a use-case:

It would be nice if I could allow a new user to:

0. Arrive at my website for the first time.
1. upload a photo (access S3)
2. execute an appsync mutation where they are the "owner" of the object. (Other users can't query this object0).
3. query appsync to list the items for which they are the "owner"
4. create an account so that they can access their resources later
5. [fifty days later, on a different computer]... login and access the photo from step 1.

Currently, I have to force them to create an account (step 4) before they can upload a photo (step 1). It would be nice if my users could test the service before being asked to create an account. Thus, It would be great if steps 1 through 3 could be performed as a guest user and then the guest user could be upgraded to an authenticated user when they perform step 4.

**Comment by pechisworks** [comment]
any news on this @powerful23 ?

**Comment by salimhamed** [comment]
@powerful23 BTW, this is still an issue for our team

**Comment by iwaduarte** [comment]
This issue was opened stunningly 9 months. 14 Interactions and lot of hacky-ways still nothing...
Started a project with Amplify and I really can see that unfortunately is not a production ready yet.

**Comment by thomers** [comment]
Another month has passed. Any update?

**Comment by davidbiller** [comment]
Any update here?

**Comment by ruodingt7** [comment]
Keen to know any update on this thread?

**Comment by davidbiller** [comment]
> Keen to know any update on this thread?

My workaround is:
Create an account with deviceID as Mail and password. Than the user can change later mail and password. 

But yes! It’s important to get this function from cognito.

**Comment by ruodingt** [comment]
@davidbiller 
Are you using Cognito for authentication?
Cognito normally requires verify email. How do you bypass that?

**Comment by davidbiller** [comment]
> @davidbiller
> Are you using Cognito for authentication?
> Cognito normally requires verify email. How do you bypass that?

@ruodingt 

Yes. I do.
Use this pre sign up lambda:

exports.handler = (event, context, callback) => {
    event.response.autoConfirmUser = true;
    context.done(null, event);
};

**Comment by ruodingt** [comment]
@davidbiller Thanks so much, that's really helpful.
So later user changes their password and email. That email has to be verified as well.
Will the verification be automatically triggered by Cognito or we have to configure somethings.

One more question may be off the topic a bit but I think it you probably also thought about it before:

Previously we use the post-confirmation trigger to send user an welcome email. However, if we automatically confirm the dummy user with only deviceID, we can't send the welcome message using post-confirmation trigger. And I didn't find other appropriate trigger for that.
Do you know a workaround for sending welcome email?

**Comment by davidbiller** [comment]
> @davidbiller Thanks so much, that's really helpful.
> So later user changes their password and email. That email has to be verified as well.
> Will the verification be automatically triggered by Cognito or we have to configure somethings.
> 
> One more question may be off the topic a bit but I think it you probably also thought about it before:
> 
> Previously we use the post-confirmation trigger to send user an welcome email. However, if we automatically confirm the dummy user with only deviceID, we can't send the welcome message using post-confirmation trigger. And I didn't find other appropriate trigger for that.
> Do you know a workaround for sending welcome email?

There is a cognito function for this. You will find it easy. 

You can’t send a mail without the mail address. But you can send it after the „fake“ signUp. 
I build a ui as a signup, but it’s a password and mail change :)

**Comment by ilyalukyanov** [comment]
Hi all, I couldn't wait and came up with a workaround. Hope it will help to progress with the proper sooner.

This TypeScript code solves it for me (should me executed before the first call to anything in Amplify):

```
import { Auth } from "aws-amplify";
import { Credentials, ICredentials, getAmplifyUserAgent } from "@aws-amplify/core";
import { CognitoIdentityClient, GetCredentialsForIdentityCommand } from '@aws-sdk/client-cognito-identity';
import { CognitoUser, IAuthenticationCallback } from 'amazon-cognito-identity-js';

type AuthCallbacksDelegate = (
  user: CognitoUser, 
  resolve: (value?: CognitoUser | any) => void, 
  reject: (value?: any) => void
) => IAuthenticationCallback;

const fixAuth = () => {
  const AuthClassInstance = Auth as any;
  const baseAuthCallbacks = AuthClassInstance.authCallbacks as AuthCallbacksDelegate;

  AuthClassInstance.authCallbacks = function(user: any, resolve: any, reject: any) {
    const callbacks = baseAuthCallbacks.call(this, user, resolve, reject);
    const baseOnSuccess = callbacks.onSuccess;
    const { region, userPoolId } = this._config;
      
    callbacks.onSuccess = async (session, userConfirmationNecessary?) => {
      const credentials = await Credentials.get() as ICredentials;
      if (credentials?.identityId && credentials.authenticated === false) {
        const cognitoClient = new CognitoIdentityClient({
          region,
          customUserAgent: getAmplifyUserAgent()
        });
      
        /* 
         * This call to GetCredentialsForIdentityCommand links the newly authenticated user with the anonymous identity in the Identity Pool.
         * When https://github.com/aws-amplify/amplify-js/blob/e11e938184fbb935374ca8f19b96365a7d6af7ce/packages/core/src/Credentials.ts#L354
         * is called further down the chain inside `baseOnSuccess`, it will return the same IdentityId instead of creating a new one.
         * Thus, the same value of UserId will be used for Pinpoint before and after registration.
         * In the case when the user is already registered and is signing in (i.e. they already have their account in User Pool linked to an Identity), 
         * this call to GetCredentialsForIdentityCommand will not make any additional links. Instead it will mark the new Identity as disabled and
         * will merge it into the old one - https://docs.aws.amazon.com/cognito/latest/developerguide/authentication-flow.html (under Merging Identities).
         * There's also this thread that covers that https://forums.aws.amazon.com/thread.jspa?threadID=238042
         */
        await cognitoClient.send(new GetCredentialsForIdentityCommand({
          IdentityId: credentials.identityId,
          Logins: {
            [`cognito-idp.${region}.amazonaws.com/${userPoolId}`]: session.getIdToken().getJwtToken()
          }
        }));
      }

      await (baseOnSuccess(session, userConfirmationNecessary) as unknown as Promise<void>);
    }  
    
    return callbacks;
  }
};
```

I guess the solution would be to just replace `GetIdCommand` here https://github.com/aws-amplify/amplify-js/blob/e11e938184fbb935374ca8f19b96365a7d6af7ce/packages/core/src/Credentials.ts#L354 with `GetCredentialsForIdentityCommand` as in the workaround.

I'm happy to send a PR, please just confirm this is the right approach.

**Comment by Maihistory** [comment]
How does GitHub Support handle and write bug reports? ... questions in your bug report is the best way to get a speedy resolution of your issue !

**Comment by SumindaD** [comment]
Any update on this issue?

**Comment by jernejc** [comment]
2 years almost ?

**Comment by NikoMontana** [comment]
Yeah, glad I decided back then to roll with my own solution.
It was a pain in the ass at first, but man - totally worth it!
As I can see other guys still struggling with this kind of shit.

AWS is not interested in AppSync, it´s just another service
in their chain, but fo real - firebase is still better even
if firebase sucks.

Am Di., 1. Sept. 2020 um 12:23 Uhr schrieb Jernej <notifications@github.com

**Comment by imshashank** [comment]
Any updates on this?

**Comment by davidbiller** [comment]
Still open?! 🤭💩

**Comment by SergeyShvager** [comment]
Any updates on this?

**Comment by ahmetcetin** [comment]
30 months and counting...

**Comment by thedgbrt** [comment]
@ahmetcetin wanna bet how long it takes from now on? I'm going for 18 months 😅

**Comment by jakemitchellxyz** [comment]
@iartemiev @elorzafe @manueliglesias idk who to tag, but y'all seem active.

I think the severity of this issue is critically underappreciated. I have an API that does not allow unauthorized credentials, but I also have another API that does. My users should be able to authenticate against the unauth role until they finally sign in, which needs to automatically switch to the auth role so that they can access the restricted API. There is nothing in the documentation about how to achieve this (probably because this is the assumed default behavior and it's just buggin'). It's a shame to see that this isn't being prioritized as a top priority. This is not a new feature, this is a critical bug in the way Amplify is handling user Identities as they change auth status.

After hours of debugging an unhelpful 403 error, I was able to use X-Ray with my API to see that my calls to my API have been using the unauth IAM role.  This is wholly incorrect behavior from Amplify because, prior to this API call, I authenticated through Cognito and all subsequent requests _should_ be using the auth role. It took forever to figure this out because I was under the impression that the auth role is used if you are logged in, but this is not the case.

![unauth_role_hitting_api](https://user-images.githubusercontent.com/13812224/115767554-e6de7200-a36e-11eb-8577-0cb50c87c79e.PNG)

I'll spare you the proof that I have everything set up correctly until y'all need something more specific.
This is just not an issue for edge cases like @ajhool 's interesting usage; this is an issue that affects any app that has one public API and one private API  (which is a looooot of cases).

Can we get some eyes on this so we can either involve the community to fix it, prioritize it in your backlog, or give us some instructions for a simple remediation?

**Comment by jakemitchellxyz** [comment]
Also, this was assigned to @powerful23 but they seem to have left the project. Possibly why this slipped through the cracks for so long.

**Comment by reesscot** [comment]
I am currently researching a solution to this issue. Thanks to everyone for posting detailed use cases.

**Comment by reesscot** [comment]
Quick update on this issue. I've tested @ilyalukyanov's [solution](https://github.com/aws-amplify/amplify-js/issues/2089#issuecomment-622331432) above, and I'm seeing that it correctly links the guest identity to the new login created at signup. One caveat to this solution is that [signOut](https://github.com/aws-amplify/amplify-js/blob/main/packages/amazon-cognito-identity-js/src/CognitoUser.js#L1969) doesn't currently remove the identity id from local storage. This causes an app to throw a 400 error when it tries to get guest AWS credentials using the identity from local storage (since the identity is now attached to an authenticated user). I'm now researching what else needs to change to support this guest->authenticated use case.

**Comment by reesscot** [comment]
@jakemitchellxyz and @salimhamed Are you still having issues with logged in users assuming the `unauth` role instead of the `auth` role? I'm unable to reproduce, and it seems like it might be a different problem than the primary ask of this issue, which is [preserving guest identity id on signup](https://github.com/aws-amplify/amplify-js/issues/2089#issuecomment-448762522). From my understanding a signed in Cognito User should always get the `auth` role. 

@jakemitchellxyz How are you setting up your API's? Are they custom APIs? If you are using the AWS AppSync GraphQL API, then your issue might be related to [2363](https://github.com/aws-amplify/amplify-js/issues/2363).

@salimhamed Are you still having issues? If so, can you please open a separate issue and provide more details with [DEBUG](https://docs.amplify.aws/lib/utilities/logger/q/platform/js#setting-logging-levels) enabled?

**Comment by reesscot** [comment]
Quick update: The PR https://github.com/aws-amplify/amplify-js/pull/8278 to remove the CognitoIdentityId from LocalStorage cache after signin/signout has been merged (but not released yet). Once that fix has been released, it should be possible to use @ilyalukyanov's [solution](https://github.com/aws-amplify/amplify-js/issues/2089#issuecomment-622331432) above as a temporarily work around for users who need guest => auth user merge functionality right away. I'm continuing to work on a plan to make this feature request available without the workaround.

**Comment by reesscot** [comment]
This issue is fixed in version 4.1.2 (see PR https://github.com/aws-amplify/amplify-js/pull/8466)
